kaiax/builder/impl/tx_pool_wrapping.go

```diff
@@ -81,37 +82,90 @@ func (b *BuilderWrappingModule) IsReady(txs map[uint64]*types.Transaction, next
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	tx, isReady := txs[next]
-	if b.txPoolModule != nil {
-		isReady = b.txPoolModule.IsReady(txs, next, ready)
+	// false if tx is not in txs
+	tx, ok := txs[next]
+	if !ok {
+		return false
+	}
+
+	// false if tx is not ready in child module
+	if b.txPoolModule != nil && !b.txPoolModule.IsReady(txs, next, ready) {
+		return false
 	}
-	if isReady && b.txBundlingModule.IsBundleTx(tx) {
-		if _, ok := b.knownTxs[tx.Hash()]; !ok {
-			b.knownTxs[tx.Hash()] = txAndTime{
-				tx:   tx,
-				time: time.Now(),
+
+	// add tx to knownTxs if it is a bundle tx and not in knownTxs
+	if b.txBundlingModule.IsBundleTx(tx) && !b.knownTxs.has(tx.Hash()) {
+		// If prev tx is bundle tx, there's no need to check the knownTxs limit because it has been checked in the previous `IsReady()` execution.
+		isPrevTxBundleTx := len(ready) != 0 && b.txBundlingModule.IsBundleTx(ready[len(ready)-1])
+		if isPrevTxBundleTx {
+			b.knownTxs.add(tx)
+			return true
+		}
+
+		maxBundleTxsInPending := b.txBundlingModule.GetMaxBundleTxsInPending()
+		if maxBundleTxsInPending != math.MaxUint64 {
+			numExecutable := uint(b.knownTxs.numExecutable())
+
+			numSeqTxs := uint(1)
+			for i := next + 1; true; i++ {
+				if _, ok := txs[i]; !ok {
+					break
+				}
+				if !b.txBundlingModule.IsBundleTx(txs[i]) {
+					break
+				}
+				numSeqTxs++
+			}
+
+			// false if there is possibility of exceeding max bundle tx num
+			if numExecutable+numSeqTxs > maxBundleTxsInPending {
+				logger.Info("Exceed max bundle tx num", "numExecutable", "maxBundleTxsInPending", maxBundleTxsInPending)
+				return false
 			}
 		}
+
+		b.knownTxs.add(tx)
 	}
-	return isReady
+
+	return true
 }
 
 // PreReset is a wrapper function that removes timed out tx from the tx pool and knownTxs.
 func (b *BuilderWrappingModule) PreReset(oldHead, newHead *types.Header) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	for hash, txAndTime := range b.knownTxs {
+	for hash, knownTx := range b.knownTxs {
 		// remove pending timed out tx from tx pool
-		if time.Since(txAndTime.time) >= PendingTimeout {
-			b.knownTxs[hash].tx.MarkUnexecutable(true)
+		if knownTx.elapsedTime() >= PendingTimeout {
+			b.knownTxs.markUnexecutable(hash)
 		}
 		// remove known timed out tx from knownTxs
-		if time.Since(txAndTime.time) >= KnownTxTimeout {
-			delete(b.knownTxs, hash)
+		if knownTx.elapsedTime() >= KnownTxTimeout {
+			b.knownTxs.delete(hash)
 		}
 	}
 	if b.txPoolModule != nil {
 		b.txPoolModule.PreReset(oldHead, newHead)
 	}
 }
+
+// PostReset is a wrapper function that calls the PostReset method of the underlying module.
+func (b *BuilderWrappingModule) PostReset(oldHead, newHead *types.Header) {
```

Comment:
Don't we need a lock?

Reply:
https://github.com/kaiachain/kaia/pull/328/commits/c7b3fe6325886423fbcd3398586230afff646fde